### PR TITLE
Support passing command line flags via `gok run`

### PR DESCRIPTION
**Pull request by markdrayton**
See https://github.com/gokrazy/gokrazy/issues/164.

Still to do/decide:

1) What locking is needed for `s.cmd.Args`? I'm tempted to just wrap a mutex around it (or `s.cmd`) but I haven't satisfied myself that I fully understand all the code and implications. I'd take a suggestion here.

2) I fiddled with making `divertRequest` public but that led to `updater` depending on `gokrazy`, which it doesn't yet. Same deal in the opposite direction if I declare it in the `updater` package. Should I put it in the `internal` repo or is what I have fine?

3) I'm assuming we don't need a fully backwards- and forwards-compatible migration path for the changes in each of the 3 affected repos and can just commit (I think) the `gokrazy` and `updater` changes first, then update `go.mod` in `tools` to point at the updated dependencies.

Test run:

```sh
# after pointing github.com/gokrazy/{gokrazy,updater} at my local repos and building an updated `gok`
$ gok -i test update
[..]

$ cd ~/go/src/github.com/gokrazy; cat hello.go
package main

import (
	"log"
	"os"
)

func main() {
	log.Printf("%+v", os.Args)
}

$ jq '.PackageConfig."github.com/gokrazy/hello"' < ~/gokrazy/test/config.json
{
  "CommandLineFlags": [
    "boop"
  ]
}

$ gok -i test run -- a b c
[done] in 0.64s
Transferred hello (2 MiB) at 4.57 MiB/s (total: 0s)
2023/02/01 23:55:26 streaming logs of service "hello" from gokrazy instance "test"
2023/02/01 23:55:26 gokrazy: attempt 0, starting ["/tmp/gok-run/hello" "boop" "a" "b" "c"]
2023/02/01 23:55:26 [/tmp/gok-run/hello boop a b c]
2023/02/01 23:55:26 gokrazy: exited successfully, stopping
^C

$ gok -i test run -- a b c d
[done] in 0.48s
Transferred hello (2 MiB) at 4.57 MiB/s (total: 0s)
2023/02/01 23:55:30 streaming logs of service "hello" from gokrazy instance "test"
2023/02/01 23:55:26 gokrazy: attempt 0, starting ["/tmp/gok-run/hello" "boop" "a" "b" "c"]
2023/02/01 23:55:26 [/tmp/gok-run/hello boop a b c]
2023/02/01 23:55:26 gokrazy: exited successfully, stopping
2023/02/01 23:55:31 gokrazy: attempt 1, starting ["/tmp/gok-run/hello" "boop" "a" "b" "c" "d"]
2023/02/01 23:55:31 [/tmp/gok-run/hello boop a b c d]
2023/02/01 23:55:31 gokrazy: exited successfully, stopping
```

image

### CODE REVIEW

**Comment by stapelberg** [comment]
Thanks for your PRs!

> 1. What locking is needed for `s.cmd.Args`? I'm tempted to just wrap a mutex around it (or `s.cmd`) but I haven't satisfied myself that I fully understand all the code and implications. I'd take a suggestion here.

Now that we modify it, we need a mutex around `s.cmd`.

> 2. I fiddled with making `divertRequest` public but that led to `updater` depending on `gokrazy`, which it doesn't yet. Same deal in the opposite direction if I declare it in the `updater` package. Should I put it in the `internal` repo or is what I have fine?

Duplicating the divertRequest is fine for now :)

> 3\. I'm assuming we don't need a fully backwards- and forwards-compatible migration path for the changes in each of the 3 affected repos and can just commit (I think) the `gokrazy` and `updater` changes first, then update `go.mod` in `tools` to point at the updated dependencies.

No, we actually do need compatibility in both directions. Some people might update their gok cli, and still expect it to work with gokrazy instances that haven’t been updated ot the latest github.com/gokrazy/gokrazy yet. On the other end of the spectrum, some folks might have module updates automated (as described on https://gokrazy.org/userguide/automation/) and will get a newer github.com/gokrazy/gokrazy, but still have an older gok cli.

In this case, it’s quite easy, though: 

1. check for the `Content-Type` header to be `application/json`. If that’s not the case, fill a divertRequest from r.FormValue() calls.
2. On the gokrazy/updater side, fall back to request parameters when you get a BadRequest HTTP response status code.

**Comment by markdrayton** [comment]
> No, we actually do need compatibility in both directions.

Gah, in hindsight that is very obvious :-(

Updated PRs coming shortly.

**Comment by markdrayton** [comment]
Okay, here are the updated PRs. I have tested this in the various combinations (new `gok`/gokrazy, old `gok`/new gokrazy, new `gok`/old gokrazy and it seems to do the job. The only thing I wondered a little about was including a warning to the user if they try to pass arguments to a program under `gok run` on a version of gokrazy that doesn't support passing them. It might be a little confusing to include arguments that appear to be ignored.

**Comment by stapelberg** [comment]
> The only thing I wondered a little about was including a warning to the user if they try to pass arguments to a program under `gok run` on a version of gokrazy that doesn't support passing them

Probably not too many users will run into it, but yeah, it would be nice. I think the best way to accomplish said error is by changing https://github.com/gokrazy/updater/pull/1 to return an error from the fallback path if len(flags) > 0

File: update.go [hunk 1]
```diff
@@ -394,23 +394,39 @@ func uploadTemp(w http.ResponseWriter, r *http.Request) {
...
 	if r.Method != http.MethodPost {
 		http.Error(w, "expected a POST request", http.StatusBadRequest)
 		return
 	}
 
-	diversion := r.FormValue("diversion")
-	if diversion == "" {
+	var req divertRequest
+	if r.Header.Get("Content-Type") == "application/json" {
+		err := json.NewDecoder(r.Body).Decode(&req)
```
Comment (stapelberg) [inline update.go:411]: This is a common mistake: json.NewDecoder “reads and decodes JSON values from an input stream.”, but we don’t have an input stream — we only have a single value. It’s better to just call json.Unmarshal() on the result of io.ReadAll(resp.Body) in this case.

**Review by stapelberg — COMMENTED** [review]
Looks good overall!

File: update.go (hunk 1 above)
Comment (markdrayton) [reply update.go]: Fixed!

**Review by stapelberg — APPROVED** [review]

**Pull request merged**